crates/viewer/re_renderer/src/renderer/video.rs

```diff
@@ -0,0 +1,544 @@
+            if timestamp.as_f64() < 0.0 {
+                // TODO(andreas): This is a hack, we should have a better way to handle this
+                return self.zeroed_texture_float.clone();
+            }
+
+            let Some(segment_idx) =
+                latest_at_idx(&self.data.segments, |segment| segment.timestamp, &timestamp)
+            else {
+                return self.texture.clone();
+            };
+
+            let Some(sample_idx) = latest_at_idx(
+                &self.data.segments[segment_idx].samples,
+                |sample| sample.timestamp,
+                &timestamp,
+            ) else {
+                // segments are never empty
+                unreachable!();
+            };
+
+            if segment_idx != self.current_segment_idx {
+                let segment_distance = segment_idx as isize - self.current_segment_idx as isize;
+                if segment_distance == 1 {
+                    // forward seek to next segment
+                    self.enqueue_all(segment_idx + 1);
+                } else {
+                    // forward seek by N>1 OR backward seek across segments
+                    self.reset();
+                    self.enqueue_all(segment_idx);
+                    self.enqueue_all(segment_idx + 1);
+                }
+            } else if sample_idx != self.current_sample_idx {
+                let sample_distance = sample_idx as isize - self.current_sample_idx as isize;
+                if sample_distance < 0 {
+                    self.reset();
+                    self.enqueue_all(segment_idx);
+                    self.enqueue_all(segment_idx + 1);
+                }
+            }
+
+            self.current_segment_idx = segment_idx;
+            self.current_sample_idx = sample_idx;
+
+            let mut frames = self.frames.lock();
+
+            let Some(frame_idx) = latest_at_idx(&frames, |(t, _)| *t, &timestamp) else {
+                // no buffered frames - texture will be blank
+                // TODO(jan): do something less bad
+                return self.texture.clone();
+            };
+            let frame = frames[frame_idx].1 .0.clone();
+
+            // drain up-to (but not including) the frame idx, clearing out any frames
+            // before it. this lets the video decoder output more frames.
+            drop(frames.drain(0..frame_idx));
+
+            let Some((frame_timestamp_ms, frame_duration_ms)) = frame
+                .timestamp()
+                .map(TimeMs::new)
+                .zip(frame.duration().map(TimeMs::new))
+            else {
+                // TODO(jan): figure out when this can happen and handle it
+                return self.texture.clone();
+            };
+
+            if TimeMs::new(timestamp.as_f64() - frame_timestamp_ms.as_f64()) > frame_duration_ms {
+                // not relevant to the user, it's an old frame.
+                return self.texture.clone();
```

Comment:
I don't follow this comment. The `frame` we looked up is outdated, so we return `self.texture`… why?

Reply:
We should return zeroed texture here. But the point is that we're not copying that frame into the texture.